datahub/dataset/pagination.py

```diff
@@ -0,0 +1,9 @@
+from rest_framework.pagination import CursorPagination
+
+
+class OMISDatasetViewCursorPagination(CursorPagination):
+    """
+    Cursor Pagination for OMISDatasetView
+    """
+
+    ordering = ('created_date', 'omis_order_reference')
```

Comment:
```suggestion
    ordering = ('created_on', 'pk')
```

It'd be good to stick with fields on the model itself (since those are easier to index) and also use `pk` rather than the order reference (as the latter is a user-facing thing, it would be better to not rely on it for cases like these).

(There is also an existing index on `('created_on', 'pk')` that it will be able to use.)

However, I can also see that that wouldn't work with how we are using `CursorPagination` in this case. I think this needs a bit more thought.

Reply:
As far as I know, db indexes is still being used on the query doesn't matter which alias we are using.

Here is the end SQL.
```
SELECT "order_order"."delivery_date", 
(SELECT STRING_AGG(U2."name", ', 'ORDER BY U2."name") AS "_annotated_value" FROM "order_order" U0 LEFT OUTER JOIN "order_order_service_types" U1 ON (U0."id" = U1."order_id") LEFT OUTER JOIN "order_servicetype" U2 ON (U1."servicetype_id" = U2."id") WHERE U0."id" = ("order_order"."id") GROUP BY U0."id") AS "services", 
"order_order"."status" AS "order_status", "order_order"."reference" AS "omis_order_reference", "order_order"."subtotal_cost" AS "net_price", "omis-invoice_invoice"."subtotal_cost" AS "subtotal", "company_company"."name" AS "company_name", "metadata_team"."name" AS "DIT_team", "metadata_country"."name" AS "market", "order_order"."created_on" AS "created_date", "order_order"."cancelled_on" AS "cancelled_date", "order_cancellationreason"."name" AS "cancellation_reason_text", "metadata_ukregion"."name" AS "UK_region", "metadata_sector"."segment" AS "sector_name", "order_order"."paid_on" AS "payment_received_date", "order_order"."completed_on" AS "completion_date", "company_contact"."first_name" AS "contact_first_name", "company_contact"."last_name" AS "contact_last_name", "company_contact"."telephone_number" AS "contact_phone_number", "company_contact"."email" AS "contact_email_address", "company_company"."address_1" AS "company_trading_address_line_1", "company_company"."address_2" AS "company_trading_address_line_2", "company_company"."address_town" AS "company_trading_address_town", "company_company"."address_county" AS "company_trading_address_county", T11."name" AS "company_trading_address_country", "company_company"."address_postcode" AS "company_trading_address_postcode", "company_company"."registered_address_1" AS "company_registered_address_line_1", "company_company"."registered_address_2" AS "company_registered_address_line_2", "company_company"."registered_address_town" AS "company_registered_address_town", "company_company"."registered_address_county" AS "company_registered_address_county", T12."name" AS "company_registered_address_country", "company_company"."registered_address_postcode" AS "company_registered_address_postcode" 

FROM "order_order" 
LEFT OUTER JOIN "omis-invoice_invoice" ON ("order_order"."invoice_id" = "omis-invoice_invoice"."id") 
INNER JOIN "company_company" ON ("order_order"."company_id" = "company_company"."id") 
LEFT OUTER JOIN "company_advisor" ON ("order_order"."created_by_id" = "company_advisor"."id") 
LEFT OUTER JOIN "metadata_team" ON ("company_advisor"."dit_team_id" = "metadata_team"."id") 
LEFT OUTER JOIN "metadata_country" ON ("order_order"."primary_market_id" = "metadata_country"."id") 
LEFT OUTER JOIN "order_cancellationreason" ON ("order_order"."cancellation_reason_id" = "order_cancellationreason"."id") 
LEFT OUTER JOIN "metadata_ukregion" ON ("order_order"."uk_region_id" = "metadata_ukregion"."id") 
LEFT OUTER JOIN "metadata_sector" ON ("order_order"."sector_id" = "metadata_sector"."id") 
INNER JOIN "company_contact" ON ("order_order"."contact_id" = "company_contact"."id") 
LEFT OUTER JOIN "metadata_country" T11 ON ("company_company"."address_country_id" = T11."id")
LEFT OUTER JOIN "metadata_country" T12 ON ("company_company"."registered_address_country_id" = T12."id") 
ORDER BY "created_date" ASC, "omis_order_reference" ASC
```

Reply:
Used aliases are not user facing. This endpoint is designed to be machine-to-machine. It's consumed by data-flow-api, which uses the response data to insert in dataworkspace. We could do this name mapping in dataworkspace but I felt like it'd be better if it's done on source data.

Reply:
I missed that `created_date` was just an alias. However, the existing index is on `('created_on', 'pk')` and so the second ordering field should be `pk`.

Regarding the aliases, I think it's confusing to have them in this repository, and it creates unnecessary coupling. If the consumer of the API wants to store the fields under different names, that's more of a decision for it (rather than a concern for this project). So I'd stick as close to the original names here as possible.

Reply:
We have to do this mapping either in datahub or dataworkspace because you don't want this mapping in datahub, I had to create a serializer just to match field names on deserialization. I don't call it coupling but now dataworkspace knows about datahub datastructure with ugly names including relations (i.e. created_by__dit_team__name).  FYI, with these changes, if one of the fields in this api endpoint is changed somehow, code change on dataworkspace is required.

Reply:
Happy to discuss further in person, but I'd add:

- in all cases API compatibility needs to be considered and fields can't just be removed
- there's a deprecation process for removing fields if it does become a necessity
- the structure is basically the same as the endpoint the Data Hub front end uses (except that it's flat rather than nested)
- we can still use aliases if we need to, the point was that they should make sense in the context of this repository